### PR TITLE
[quant] ConvTranspose1d / ConvTranspose2d

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43199 [quant] conv_transpose graph patterns
* **#40371 [quant] ConvTranspose1d / ConvTranspose2d**
* #40370 [quant] conv_transpose1d / conv_transpose2d

Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

### CODE REVIEW

**Comment by z-a-f** [comment]
FYI: https://github.com/pytorch/pytorch/pull/36790

**Review by vkuzo — COMMENTED** [review]
would be good to add test coverage, we can review more in detail once there is more evidence on the memory corruption bug

File: torch/nn/quantized/modules/conv.py [hunk 1]
```diff
@@ -45,8 +45,12 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
         self.groups = groups
         self.padding_mode = padding_mode
         # Initialize as NCHW. set_weight will internally transpose to NHWC.
+        if self.transposed:
+            weight_shape = [in_channels, out_channels // self.groups]
```
Comment (kimishpatel) [inline torch/nn/quantized/modules/conv.py:49]: I suspect something wrong here. Are you sure dimensions are supposed to be like this. I see below that `bias_float`'s size is out_channels. It should also depend on transposed param? So here : https://github.com/pytorch/pytorch/pull/40370/files#diff-d0713fbda560de6aecef93e4c3d973a5R561, you are providing #of output channels differently than what your weight is telling you.
But more importantly what is going on with I dont understand why in deconv weigth's #of output channels is equal to in_channel? I would have thought it should not change right? Your deconv op's weight should tell you how many output channels you want to produce and that should not depend on # of input channels. No? Only thing different here I imagine is output shape calculation.

Comment (z-a-f) [reply torch/nn/quantized/modules/conv.py:49]: bias size should always be equal to the number of output channels. The shape of the weight matrix in convolution and transposed convolution is different:

- convolution weight shape: [out_channels, in_channels / groups, k0, k1, k2, ...]
- convolution transpose weight shape: [in_channels, out_channels / groups, k0, k1, k2, ...]

Can you elaborate what you mean when you say "deconv weights #of output channels equals to in_channels"?

**Comment by z-a-f** [comment]
> would be good to add test coverage, we can review more in detail once there is more evidence on the memory corruption bug

The unittests are there, part of the previous PRs

File: torch/nn/quantized/modules/conv.py (hunk 1 above)
Comment (z-a-f) [reply torch/nn/quantized/modules/conv.py:49]: Thanks Kimish for finding the bug!

File: torch/nn/quantized/modules/conv.py [hunk 2]
```diff
@@ -508,3 +509,194 @@ def _input_padding(self, kernel_size, dilation, padding):
...
+    _FLOAT_MODULE = nn.ConvTranspose1d
+
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, output_padding=0, groups=1, bias=True,
+                 dilation=1, padding_mode='zeros'):
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+        padding = _pair(padding)
+        dilation = _pair(dilation)
+        output_padding = _pair(output_padding)
```
Comment (vkuzo) [inline torch/nn/quantized/modules/conv.py:584]: can these live on the base class?

File: torch/nn/quantized/modules/conv.py [hunk 3]
```diff
@@ -508,3 +509,194 @@ def _input_padding(self, kernel_size, dilation, padding):
...
+        w, b = torch.ops.quantized.conv_transpose1d_unpack(self._packed_params)
+        return w, b
+
+    def weight(self):
+        (w, _) = self._weight_bias()
+        return w
+
+    def bias(self):
+        (_, b) = self._weight_bias()
+        return b
```
Comment (vkuzo) [inline torch/nn/quantized/modules/conv.py:612]: and these on the base class?

**Review by vkuzo — COMMENTED** [review]
looks good.  should we add a test?

**Comment by neuromorpheus** [comment]
Hi,

When is this feature expected to be available in pytorch?

**Comment by z-a-f** [comment]
> Hi,
> 
> When is this feature expected to be available in pytorch?

Hopefully soon. You mind I ask where you use it?

**Comment by neuromorpheus** [comment]
I want to use it for upsampling quantized tensors. I know I can use
interpolate, but I wanted to compare timings with a transposed convolution.

File: torch/nn/quantized/modules/conv.py (hunk 2 above)
Comment (z-a-f) [reply torch/nn/quantized/modules/conv.py:565]: I don't think so -- these are supposed to be different for every spatial dim (`_single`, `_pair`, `_triple`, etc.). The reason we are reusing because it's a "hack" to utilize the 2d to do the 1d computation. However, in the future this will change. If we put it in the base class, this might propagate through. You can take a look at the `conv.py` in the non-quantized space -- that's the pattern I am aiming for.

File: torch/nn/quantized/modules/conv.py (hunk 3 above)
Comment (z-a-f) [reply torch/nn/quantized/modules/conv.py:590]: Sure will do -- I guess we should add tests for the ConvXd modules as well.

**Review by vkuzo — APPROVED** [review]
lg as long as CI passes.  Can we include the context and the test plan in the PR summary before landing, since this is non-trivial?
